icechunk/src/repository.rs

```diff
@@ -424,9 +424,24 @@ impl Repository {
         coord: ChunkIndices,
         data: Option<ChunkPayload>,
     ) -> RepositoryResult<()> {
-        self.get_array(&path)
-            .await
-            .map(|node| self.change_set.set_chunk_ref(node.id, coord, data))
+
+        let node_snapshot = self.get_array(&path).await?;
+
+        if let NodeData::Array(zarr_metadata, _, ) = node_snapshot.node_data {
+            if zarr_metadata.valid_chunk_coord(&coord) {
+                self.change_set.set_chunk_ref(node_snapshot.id, coord, data);
+                Ok(())
+            } else {
+                Err(RepositoryError::FormatError(IcechunkFormatError::ChunkCoordinatesNotFound { coords: coord }))
+            }
+                     
+        } else {
+            Err(RepositoryError::NotAnArray {
+                node: node_snapshot,
+                message: "getting an array".to_string(),
+            })
+        }
+        
```

Comment:
maybe we can add a test for `Repository` too? that validates that the error is correctly bubbling up.